contracts/multisig-prover/src/execute.rs

```diff
@@ -0,0 +1,161 @@
+use cosmwasm_std::{
+    to_binary, wasm_execute, BlockInfo, DepsMut, Env, HexBinary, QuerierWrapper, QueryRequest,
+    Response, SubMsg, WasmMsg, WasmQuery,
+};
+
+use std::{collections::HashMap, str::FromStr};
+
+use axelar_wasm_std::{Participant, Snapshot};
+use connection_router::{msg::Message, types::ChainName};
+use service_registry::state::Worker;
+
+use crate::{
+    contract::START_MULTISIG_REPLY_ID,
+    error::ContractError,
+    events::Event,
+    state::{Config, COMMANDS_BATCH, CONFIG, KEY_ID, REPLY_BATCH},
+    types::{BatchID, CommandBatch},
+};
+
+pub fn construct_proof(deps: DepsMut, message_ids: Vec<String>) -> Result<Response, ContractError> {
+    let key_id = KEY_ID.load(deps.storage)?;
+    let config = CONFIG.load(deps.storage)?;
+
+    let batch_id = BatchID::new(&message_ids);
+
+    let messages = get_messages(
+        deps.querier,
+        message_ids,
+        config.gateway.into(),
+        config.chain_name,
+    )?;
+
+    let command_batch = match COMMANDS_BATCH.may_load(deps.storage, &batch_id)? {
+        Some(batch) => batch,
+        None => {
+            let batch = CommandBatch::new(messages, config.destination_chain_id)?;
+
+            COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+
+            batch
+        }
+    };
+
+    // keep track of the batch id to use during submessage reply
+    REPLY_BATCH.save(deps.storage, &command_batch.id)?;
+
+    let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+        key_id,
+        msg: command_batch.msg_to_sign(),
+    };
+
+    let wasm_msg = wasm_execute(config.multisig, &start_sig_msg, vec![])?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_success(wasm_msg, START_MULTISIG_REPLY_ID)))
+}
+
+fn get_messages(
+    querier: QuerierWrapper,
+    message_ids: Vec<String>,
+    gateway: String,
+    chain_name: ChainName,
+) -> Result<Vec<Message>, ContractError> {
+    let length = message_ids.len();
+
+    let query = gateway::msg::QueryMsg::GetMessages { message_ids };
+    let messages: Vec<Message> = querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+        contract_addr: gateway,
+        msg: to_binary(&query)?,
+    }))?;
+
+    if messages.len() != length {
```

Comment:
see @cjcobb23 comment here: https://github.com/axelarnetwork/axelar-amplifier/pull/50#discussion_r1303070842. Maybe we should change it to panic, if some time in the future we change the gateway query behavior to something else, we should probably review the prover as well to decide if something needs to be changed as well. A panic will ensure we pay attention to this part.

Reply:
Yes, then it should definitely panic IMO. Also, if you don't want to rely on any semantics of the gateway, then the code here should also check if `messages` are all distinct and contain all the message IDs given. @cjcobb23

Reply:
Eh, I think we can assume the gateway is not going to return duplicates or random messages that you didn't ask for. I think those would be legitimate bugs in the gateway. It doesn't really hurt to add checks for that, but I don't feel its totally necessary.
 
We definitely shouldn't assume order is the same, but I'm not sure if the code currently relies on the order actually.

Reply:
from my point of view, the order is only relevant for deriving the batch ID, but the prover doesn't rely on the order of messages received, it sorts them itself

Reply:
I'm approving the PR, but I don't like us making assumptions about something and being extra cautious about others. This to me, is always a failure of the API design. Either it should just be get_message so that we no longer worry about it all together, or get_messages should return Vec<Option<Message>>.